src/main/java/org/broadinstitute/hellbender/engine/ReferenceContext.java

```diff
@@ -425,4 +425,8 @@ public String getKmerAround(final int center, final int numBasesOnEachSide){
 
         return new String(getBases(newWindow));
     }
+
+    public void close (){
+        dataSource.close();
+    }
```

Comment:
These context objects are not meant to be closeable. Instead, the engine takes care of closing the underlying datasource at tool shutdown.

Does this patch actually resolve your "too many open files" error in practice?

Reply:
It helped since the processing of my BAM file that locally (block size 32MB) consists > 500 partitions is now failing after processing ~400 partitions instead of ~30. But it still does not fix it. That was my first attempt. But the problem I think is that with ReferenceFileSource that it leaves open file descriptors and I guess the problem is only replicable if you process larger files locally. Not sure if this might be also related but ReferenceFileSource implements AutoClosable but it isn't surrounded with try-with-resources block: https://github.com/broadinstitute/gatk/blob/c918f8522f2bd5103509b7c04d4b2f7d3fa3eb1e/src/main/java/org/broadinstitute/hellbender/engine/spark/LocusWalkerSpark.java#L113 I tried to add it but then it start failing from the very beginning of the processing.

Reply:
@droazen could you please give me any hint so I can try to dig deeper to find the fix?

Reply:
@droazen could you pls somehow help me so that I could work on the fix ?

Reply:
@mwiewior We are still looking into this one, but I believe at this point @lbergelson has some ideas as to what might be going on. @lbergelson would you mind posting your thoughts on Spark file handle leakage here?

Reply:
Ok, thanks Guys @lbergelson @droazen  - if there is anything I can help with just ping me !

Reply:
@mwiewior I see what you're aiming at with this PR.  Each shard is creating a new reader that isn't getting closed.  I don't think making the context closeable makes sense though because in most of the codebase we would never want to close that (we typicall re-use the underlying datasource many times).  I think we need to restructure the map functions so the underlying reader is closed after the operations are complete instead of making the context closeable.

Reply:
@lbergelson @droazen  I think I found the root cause of the problem - I tried to  rewrite the code into combination of mapPartitions and flatMap so that one new reader is created per partition not per each shard? Could please take a look at it ? It solved the issue for me.